library_management_system_downloader/run_api_download_and_parse_results.py

```diff
@@ -0,0 +1,201 @@
+import logging
+import pandas as pd
+from sqlalchemy import create_engine
+from sqlalchemy import event
+from sqlalchemy import Table, MetaData, Column, String, DateTime, Integer, func
+from sqlalchemy import ForeignKey, distinct
+from sqlalchemy.orm import sessionmaker
+
+# =============================================================================
+# Load configuration variables
+# =============================================================================
+
+# Location of configuration file.
+# This should be a python script (named with a .py file extension), and
+# should contain  variables named the following:
+# - static_parameters_for_api (a dictionary)
+# - user_agent_custom_string (a string)
+# - api_base_url (a string)
+# -tsv_dataset_location (a filepath string)
+
+from library_management_system_downloader \
+        import downloader_configuration_file as config
+# config.api_base_url
+# config.user_agent_custom_string
+# config.static_parameters_for_api
+# config.tsv_dataset_location
+
+from library_management_system_downloader import create_api_request as api
+from library_management_system_downloader \
+        import evaluate_api_response_for_fulltext_indication as fulltext
+
+# =============================================================================
+# Create additional settings
+# =============================================================================
+
+logging.basicConfig(level=logging.INFO)
+
+# =============================================================================
+# Set up a connection to (and tables within) an SQLite database:
+# =============================================================================
+
+# Initialize the database:
+sql_engine = create_engine(
+        'sqlite:///library_coverage_xml_and_fulltext_indicators.db')
+
+
+# From the sqlalchemy documentation, at
+# http://docs.sqlalchemy.org/en/latest/dialects/sqlite.html:
+# Turn on write-ahead logging, which will allow reading and writing at
+# the same time (without this, I was getting an error re: the database being
+# locked when chunksize is used with read_sql_query, since that returns a
+# generator and is thus still reading the database:
+@event.listens_for(sql_engine, "connect")
+def set_sqlite_pragma(sql_engine, connection_record):
+    cursor = sql_engine.cursor()
+    cursor.execute("PRAGMA journal_mode=WAL")
+    cursor.close()
+
+# Create an sqlalchemy metadata schema:
+metadata = MetaData(sql_engine)
+
+# Create a schema for the table that will hold the API data:
+# The full_text_indicator column will be an integer in that 1 = fulltext
+# available and 0 = fulltext not available:
+dois_table = Table(
```

Comment:
I think we should get rid of this table and just index the doi column of `library_holdings_table`.

Reply:
Just noting that I responded re: this table below, [here](https://github.com/greenelab/library-access/pull/7/files/2e939e44d9b92ae15f1b313248830d6c96895720#diff-de52c0e6143e6b9a01ab5635c6edbc40). We can certainly go back to a single-table setup, but I think that the two-table setup does have advantages (including not repeating DOI strings).

Reply:
> including not repeating DOI strings

What's the problem with repeating DOI strings? The extra disk space burden is irrelevant compared to the size the of the XML responses. IMO creating an internal set of identifiers is much more likely to cause a problem.

Reply:
There's no problem with repeating DOI strings, nor is there, I think, with not repeating them. If repeat downloads are to be allowed (which isn't something we talked about originally, but does make sense to me and makes development easier, as I mentioned in a [comment above](https://github.com/greenelab/library-access/pull/7#discussion_r145502339) (linking to it because there are a lot of comments in this thread at this point), DOI can't be the Primary Key of the table. In that case, we could have one of the following:

1. A table with a throwaway Primary Key and an indexed DOI column: This isn't *wrong;* it just violates [convention](https://en.wikipedia.org/wiki/First_normal_form), so I didn't use it here.
2. A table without a Primary Key: As above, not *wrong* (whatever that would mean here), just [unconventional](https://stackoverflow.com/questions/840162/should-each-and-every-table-have-a-primary-key), and potentially making it harder for others to re-tool the work for any expansions they want to do. My understanding from your [comment above](https://github.com/greenelab/library-access/pull/7#discussion_r146052018) is that this is what you're suggesting; is that correct?

Having thought on this, I don't think that there's a "correct" answer -- the current implementation works, and, to my understanding, follows some schema best practices. Having said that, I don't disagree with what you've written; I see it as basically an aesthetic choice. The design works now and would also work the other way. With that in mind, my goal is to have this PR accepted, which I know is at your discretion. As the project lead, do you have strong feelings about this? In suggesting the one-table solution, are you *requiring* the one-table solution?

Reply:
(I hope that that comment comes across as fully respectfully as I mean for it to; I think I'm just confused about where you're requiring changes vs. just suggesting them / opening conversation about design decisions in your comments.)

Reply:
> where you're requiring changes vs. just suggesting them / opening conversation about design decisions in your comments

All the comments are suggestions. If you disagree then let me know. Of course, sometimes I'll make comments that are wrong. The current level of discussion is good.

Regarding the database design (one v two tables), I'll leave it up to you. I was mostly envisioning the database as an indexed on-disk way to store the API queries. But if you're envisioning doing more with the DB than a more robust design may make sense.

Reply:
Following your comment [above](https://github.com/greenelab/library-access/pull/7#discussion_r147771042),

> Regarding the database design (one v two tables), I'll leave it up to you. I was mostly envisioning the database as an indexed on-disk way to store the API queries. But if you're envisioning doing more with the DB than a more robust design may make sense.

I do take your point, and think that ["You Aren't Gonna Need It"](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) is generally a worthwhile philosophy to adhere to. In this particular case, I do prefer the two-table design. Not because I think the one-table design is worse, but because I can imagine myself playing with the data in a way that would make the two-table design useful. For example, if we did bump this up to querying millions of DOIs, as you originally suggested, it seems to me that it could be useful in that case to import the State of OA dataset (with the OA "color" column) into that second table, to allow faster subsetting of DOI results based on color. That's something that we don't need to do right now, but could easily want to do. And having the second table in place would make that quicker to do.

Had I had "YAGNI" in mind, I probably wouldn't have made the second table. But since I have, and it's a quick `JOIN` to get it into a single table, it makes more sense to me to leave it with two tables. Does that sound reasonable to you, too?